### PR TITLE
Honour Insertion order for listeners

### LINKED ISSUE #2558: Make IExecutionListener, ITestListener, IInvokedMethodListener, IConfigurationListener, ISuiteListenerexecute in the order of insertion
### Expected behavior
If multiple listeners implements the same ITestNGListener, like IExecutionListener, ITestListener, IInvokedMethodListener IConfigurationListener, or ISuiteListenerexecute,  those listeners should be executed as inserted order when **Start etc , meanwhile reverse order when **Finish/Failure etc.

### Actual behavior
Listeners implments the same ITestNGListener are executed inorderred.

### Is the issue reproductible on runner?

All

### PR DESCRIPTION
**Pull request by krmahadevan**
Closes #2558

Fixes #2558  .

### Did you remember to?

- [X] Add test case(s)
- [X] Update `CHANGES.txt`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

### CODE REVIEW

File: testng-core/src/main/java/org/testng/SuiteRunner.java [hunk 1]
```diff
@@ -37,15 +37,15 @@
       Collections.synchronizedMap(Maps.newLinkedHashMap());
   private final List<TestRunner> testRunners = Lists.newArrayList();
   private final Map<Class<? extends ISuiteListener>, ISuiteListener> listeners =
-      Maps.newConcurrentMap();
+      Maps.newLinkedHashMap();
```
Comment (juherr) [inline testng-core/src/main/java/org/testng/SuiteRunner.java:40]: Why not synchronizedLinkedHashMap?

File: testng-core/src/main/java/org/testng/SuiteRunner.java [hunk 2]
```diff
@@ -189,25 +192,20 @@ private void init(
...
     }
     // Add our own IInvokedMethodListener
-    invokedMethodListeners = Maps.newConcurrentMap();
-    if (invokedMethodListener != null) {
-      for (IInvokedMethodListener listener : invokedMethodListener) {
-        invokedMethodListeners.put(listener.getClass(), listener);
-      }
+    invokedMethodListeners = Maps.synchronizedLinkedHashMap();
+    for (IInvokedMethodListener listener :
+        Optional.ofNullable(invokedMethodListener).orElse(Collections.emptyList())) {
```
Comment (juherr) [inline testng-core/src/main/java/org/testng/SuiteRunner.java:197]: invokedMethodListener is not nullable, it was just created one line before.

File: testng-core/src/main/java/org/testng/SuiteRunner.java [hunk 3]
```diff
@@ -254,7 +252,7 @@ public void setReportResults(boolean reportResults) {
   }
 
   private void invokeListeners(boolean start) {
-    for (ISuiteListener sl : listeners.values()) {
+    for (ISuiteListener sl : Lists.newArrayList(listeners.values())) {
```
Comment (juherr) [inline testng-core/src/main/java/org/testng/SuiteRunner.java:255]: Using a new ArrayList here won't change things

File: testng-core/src/test/java/test/listeners/github2558/TestClassSamples.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
+package test.listeners.github2558;
+
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+public class TestClassSamples {
+
+  //  @Listeners({
```
Comment (juherr) [inline testng-core/src/test/java/test/listeners/github2558/TestClassSamples.java:9]: Wy is it commented? Document the reason or remove.

File: testng-core/src/main/java/org/testng/SuiteRunner.java (hunk 2 above)
Comment (krmahadevan) [reply testng-core/src/main/java/org/testng/SuiteRunner.java:197]: Not true. Please check the variable names. We are doing a null check on `invokedMethodListener` using `Optional.ofNullable()`, but we are initialising a different variable `invokedMethodListeners` (this is a plural)

File: testng-core/src/test/java/test/listeners/github2558/TestClassSamples.java (hunk 4 above)
Comment (krmahadevan) [reply testng-core/src/test/java/test/listeners/github2558/TestClassSamples.java:9]: That was not supposed to be there. Removed it.

File: testng-core/src/main/java/org/testng/SuiteRunner.java (hunk 3 above)
Comment (krmahadevan) [reply testng-core/src/main/java/org/testng/SuiteRunner.java:255]: Yes they would.  This was done to accommodate the tests that are part of https://github.com/cbeust/testng/issues/2061

File: testng-core/src/main/java/org/testng/SuiteRunner.java (hunk 1 above)
Comment (krmahadevan) [reply testng-core/src/main/java/org/testng/SuiteRunner.java:40]: I dont see a concurrent usage in this case. That is why I didnt use the synchronized variant.

**Comment by krmahadevan** [comment]
@juherr - Please help check again.

File: testng-core/src/main/java/org/testng/SuiteRunner.java (hunk 3 above)
Comment (juherr) [reply testng-core/src/main/java/org/testng/SuiteRunner.java:255]: How is it possible? The order of the list will depend on the order of `listeners.values()` and I can't remember `Lists.newArrayList` make a sort.

File: testng-core/src/main/java/org/testng/SuiteRunner.java (hunk 2 above)
Comment (juherr) [reply testng-core/src/main/java/org/testng/SuiteRunner.java:197]: Oops, right! :)

File: testng-core/src/main/java/org/testng/SuiteRunner.java (hunk 1 above)
Comment (juherr) [reply testng-core/src/main/java/org/testng/SuiteRunner.java:40]: If it doesn't break #2063, then ok

**Review by juherr — APPROVED** [review]

File: testng-core/src/main/java/org/testng/SuiteRunner.java (hunk 3 above)
Comment (krmahadevan) [reply testng-core/src/main/java/org/testng/SuiteRunner.java:255]: @juherr  its insertion order remember ? So the order in which `listener.values()` provides is the way in which they would get added to the new list as well.

Comment (juherr) [reply testng-core/src/main/java/org/testng/SuiteRunner.java:255]: I think `listeners.values()` and `Lists.newArrayList(listeners.values())` will both produce the same same Iterator.

**Pull request merged**